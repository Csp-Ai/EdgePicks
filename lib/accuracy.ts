import { getSupabaseClient } from './supabaseClient';
import { agents as agentRegistry } from './agents/registry';
import type { AgentName, AgentOutputs } from './types';

interface MatchupRow {
  agents: AgentOutputs;
  pick: { winner?: string } | null;
  actual_winner: string | null;
  flow: string | null;
}

export interface AccuracyStat {
  name: string;
  wins: number;
  losses: number;
  accuracy: number;
}

export async function recomputeAccuracy() {
  const client = getSupabaseClient();
  const { data, error } = await client
    .from('matchups')
    .select('agents, pick, actual_winner, flow');
  if (error || !data) {
    throw new Error(error?.message || 'Failed to fetch matchups');
  }

  const agentTallies: Record<AgentName, { wins: number; losses: number }> = {};
  agentRegistry.forEach(({ name }) => {
    agentTallies[name as AgentName] = { wins: 0, losses: 0 };
  });
  const flowTallies: Record<string, { wins: number; losses: number }> = {};

  data.forEach((row: MatchupRow) => {
    const actual = row.actual_winner;
    if (!actual) return;

    // Agent stats
    agentRegistry.forEach(({ name }) => {
      const pick = row.agents?.[name]?.team;
      if (!pick) return;
      if (pick === actual) {
        agentTallies[name as AgentName].wins += 1;
      } else {
        agentTallies[name as AgentName].losses += 1;
      }
    });

    // Flow stats
    const flowName = row.flow || 'unknown';
    const predicted = row.pick?.winner;
    if (!flowTallies[flowName]) flowTallies[flowName] = { wins: 0, losses: 0 };
    if (predicted) {
      if (predicted === actual) {
        flowTallies[flowName].wins += 1;
      } else {
        flowTallies[flowName].losses += 1;
      }
    }
  });

  const agentStats = Object.entries(agentTallies).map(([agent, t]) => ({
    agent,
    wins: t.wins,
    losses: t.losses,
    accuracy: t.wins + t.losses > 0 ? t.wins / (t.wins + t.losses) : 0,
  }));

  const flowStats = Object.entries(flowTallies).map(([flow, t]) => ({
    flow,
    wins: t.wins,
    losses: t.losses,
    accuracy: t.wins + t.losses > 0 ? t.wins / (t.wins + t.losses) : 0,
  }));

  await client.from('agent_stats').upsert(agentStats, { onConflict: 'agent' });
  await client.from('flow_stats').upsert(flowStats, { onConflict: 'flow' });

  return { agentStats, flowStats };
}

