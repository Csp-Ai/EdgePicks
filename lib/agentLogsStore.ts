export interface AgentLog {
  output?: any;
  durationMs?: number;
  error?: string;
}

const agentLogs: Record<string, AgentLog> = {};

export function writeAgentLog(
  sessionId: string,
  agentId: string,
  data: AgentLog
) {
  agentLogs[`${sessionId}:${agentId}`] = data;
}

export function readAgentLog(sessionId: string, agentId: string): AgentLog | undefined {
  return agentLogs[`${sessionId}:${agentId}`];
}
