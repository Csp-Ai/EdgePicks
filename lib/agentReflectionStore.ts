import fs from 'fs';
import path from 'path';

export interface AgentReflection {
  whatIObserved: string;
  whatIChose: string;
  whatCouldImprove: string;
}

const filePath = path.join(process.cwd(), 'agent-reflections.json');

export function writeAgentReflection(agent: string, reflection: AgentReflection) {
  let data: Record<string, AgentReflection> = {};
  try {
    data = JSON.parse(fs.readFileSync(filePath, 'utf8'));
  } catch {
    // ignore
  }
  data[agent] = reflection;
  fs.writeFileSync(filePath, JSON.stringify(data, null, 2));
}

export function readAgentReflections(): Record<string, AgentReflection> {
  try {
    return JSON.parse(fs.readFileSync(filePath, 'utf8')) as Record<string, AgentReflection>;
  } catch {
    return {};
  }
}
