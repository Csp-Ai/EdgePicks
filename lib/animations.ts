export const FADE_DURATION = 0.4;
export const EASE = 'easeInOut';
