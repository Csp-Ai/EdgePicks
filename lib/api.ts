function cleanHeaders(input?: HeadersInit): HeadersInit | undefined {
  if (!input) return undefined;
  if (Array.isArray(input)) return input.filter(([, v]) => v !== undefined) as [string, string][];
  if (input instanceof Headers) return input;
  const out: Record<string, string> = {};
  for (const [k, v] of Object.entries(input)) {
    if (typeof v === "string") out[k] = v;
  }
  return out;
}

export async function apiGet<T>(url: string, init?: RequestInit): Promise<T> {
  const headers = cleanHeaders(init?.headers);
  const res = await fetch(url, { ...init, headers });
  if (!res.ok) throw new Error(`${res.status} ${res.statusText}`);
  return res.json() as Promise<T>;
}

// Convenience typed endpoints
import type { UpcomingGame, AgentEvent, Prediction } from './types';

export const getUpcomingGames = () => apiGet<UpcomingGame[]>('/api/upcoming-games');
export const getAgentEvents = () => apiGet<AgentEvent[]>('/api/agent-events');
export const getPrediction = (gameId: string) =>
  apiGet<Prediction>(`/api/run-predictions?gameId=${encodeURIComponent(gameId)}`);
