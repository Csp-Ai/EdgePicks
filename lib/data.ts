import { ENV } from './env';

export interface Game {
  homeTeam: string;
  awayTeam: string;
  league: string;
  time: string;
  confidence: number;
  edgePick: any[];
  winner: string;
  edgeDelta: number;
  odds?: {
    spread?: number;
    overUnder?: number;
    moneyline?: { home?: number; away?: number };
    bookmaker?: string;
    lastUpdate?: string;
  };
}

export async function fetchUpcomingGames(league?: string): Promise<Game[]> {
  try {
    const response = await fetch(
      `${ENV.NEXT_PUBLIC_SITE_URL}/api/upcoming-games${league ? `?league=${league}` : ''}`,
      {
        next: { revalidate: 300 } // Cache for 5 minutes
      }
    );

    if (!response.ok) {
      throw new Error('Failed to fetch upcoming games');
    }

    const games = await response.json();
    
    if (league) {
      return games.filter((game: Game) => game.league === league);
    }

    return games;
  } catch (error) {
    console.error('Error fetching upcoming games:', error);
    return [];
  }
}

export async function fetchGamePrediction(homeTeam: string, awayTeam: string): Promise<Game | null> {
  try {
    const response = await fetch(
      `${ENV.NEXT_PUBLIC_SITE_URL}/api/run-predictions`,
      {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json'
        },
        body: JSON.stringify({
          homeTeam,
          awayTeam
        })
      }
    );

    if (!response.ok) {
      throw new Error('Failed to fetch game prediction');
    }

    return response.json();
  } catch (error) {
    console.error('Error fetching game prediction:', error);
    return null;
  }
}
