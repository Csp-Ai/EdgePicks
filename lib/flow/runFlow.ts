import type {
  Matchup,
  AgentOutputs,
  AgentResult,
  AgentName,
  AgentLifecycle,
} from '../types';
import { loadAgents } from '../agents/loadAgents';
import type { FlowConfig } from './loadFlow';
import pLimit from 'p-limit';
import { ENV } from '../env';
import crypto from 'crypto';
import { getCacheDriver } from '../infra/cache';

export interface AgentExecution {
  name: AgentName;
  result?: AgentResult;
  error?: true;
  errorInfo?: { message?: string; stack?: string };
  scoreTotal?: number;
  confidenceEstimate?: number;
  agentDurationMs?: number;
  sessionId?: string;
}

export interface FlowRunResult {
  outputs: Partial<AgentOutputs>;
  executions: AgentExecution[];
}

function cacheKey(flow: FlowConfig, matchup: Matchup) {
  const version = ENV.FLOW_CACHE_VERSION;
  const inputHash = crypto
    .createHash('sha1')
    .update(JSON.stringify(matchup))
    .digest('hex');
  const agentsHash = crypto
    .createHash('sha1')
    .update(flow.agents.join(','))
    .digest('hex');
  return `flow:${flow.name}:${version}:${agentsHash}:${inputHash}`;
}

export async function runFlow(
  flow: FlowConfig,
  matchup: Matchup,
  onAgent?: (exec: AgentExecution) => void,
  onLifecycle?: (event: { name: AgentName } & AgentLifecycle) => void
): Promise<FlowRunResult> {
  const key = cacheKey(flow, matchup);
  const cacheDriver = getCacheDriver();
  const cached = await cacheDriver.get<FlowRunResult>(key);
  if (cached) {
    console.log('cache:hit');
    return cached;
  }

  const outputs: Partial<AgentOutputs> = {};
  const executions: AgentExecution[] = new Array(flow.agents.length);
  const agents = await loadAgents();
  const getAgent = (name: AgentName) => agents.find((a) => a.name === name);
  const limit = pLimit(ENV.MAX_FLOW_CONCURRENCY);

  let index = 0;
  while (index < flow.agents.length) {
    const batch: { name: AgentName; idx: number; agent: typeof agents[number] }[] = [];

    while (index < flow.agents.length) {
      const name = flow.agents[index];
      const agent = getAgent(name);
      if (!agent) {
        console.error(`[runFlow] Agent not found: ${name}`);
        const exec: AgentExecution = { name, error: true };
        executions[index] = exec;
        onAgent?.(exec);
        index++;
        continue;
      }
      if (agent.run.length >= 2) break;
      batch.push({ name, idx: index, agent });
      index++;
    }

    if (batch.length > 0) {
      await Promise.allSettled(
        batch.map(({ name, idx, agent }) =>
          limit(async () => {
            console.log(`[runFlow] ${name} input:`, matchup);
            const start = Date.now();
            onLifecycle?.({ name, status: 'started', startedAt: start });
            try {
              const result = await agent.run(matchup, outputs);
              const end = Date.now();
              const duration = end - start;
              console.log(`[runFlow] ${name} output:`, result);
              outputs[name] = result;
              const exec: AgentExecution = { name, result };
              executions[idx] = exec;
              onAgent?.(exec);
              onLifecycle?.({
                name,
                status: 'completed',
                startedAt: start,
                endedAt: end,
                durationMs: duration,
              });
            } catch (err: any) {
              const end = Date.now();
              const duration = end - start;
              console.error(`[runFlow] ${name} error:`, err);
              const errorInfo = { message: err?.message || 'Unknown error', stack: err?.stack };
              const exec: AgentExecution = { name, error: true, errorInfo };
              executions[idx] = exec;
              onAgent?.(exec);
              onLifecycle?.({
                name,
                status: 'errored',
                startedAt: start,
                endedAt: end,
                durationMs: duration,
                error: errorInfo,
              });
            }
          })
        )
      );
    }

    if (index < flow.agents.length) {
      const name = flow.agents[index];
      const agent = getAgent(name);
      if (!agent) {
        console.error(`[runFlow] Agent not found: ${name}`);
        const exec: AgentExecution = { name, error: true };
        executions[index] = exec;
        onAgent?.(exec);
        index++;
        continue;
      }
      console.log(`[runFlow] ${name} input:`, matchup);
      const start = Date.now();
      onLifecycle?.({ name, status: 'started', startedAt: start });
      try {
        const result = await agent.run(matchup, outputs);
        const end = Date.now();
        const duration = end - start;
        console.log(`[runFlow] ${name} output:`, result);
        outputs[name] = result;
        const exec: AgentExecution = { name, result };
        executions[index] = exec;
        onAgent?.(exec);
        onLifecycle?.({
          name,
          status: 'completed',
          startedAt: start,
          endedAt: end,
          durationMs: duration,
        });
      } catch (err: any) {
        const end = Date.now();
        const duration = end - start;
        console.error(`[runFlow] ${name} error:`, err);
        const errorInfo = { message: err?.message || 'Unknown error', stack: err?.stack };
        const exec: AgentExecution = { name, error: true, errorInfo };
        executions[index] = exec;
        onAgent?.(exec);
        onLifecycle?.({
          name,
          status: 'errored',
          startedAt: start,
          endedAt: end,
          durationMs: duration,
          error: errorInfo,
        });
      }
      index++;
    }
  }

  const result: FlowRunResult = { outputs, executions };
  await cacheDriver.set(key, result, ENV.PREDICTION_CACHE_TTL_SEC);
  return result;
}

