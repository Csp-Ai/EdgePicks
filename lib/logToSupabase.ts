import { getSupabaseClient } from './supabaseClient';
import { AgentOutputs, Matchup, PickSummary } from './types';

export async function logToSupabase(
  matchup: Matchup,
  agents: AgentOutputs,
  pick: PickSummary,
  loggedAt: string = new Date().toISOString()
): Promise<string> {
  const client = getSupabaseClient();
  const { error } = await client.from('matchups').insert({
    team_a: matchup.homeTeam,
    team_b: matchup.awayTeam,
    match_day: matchup.matchDay,
    agents,
    pick,
    created_at: loggedAt,
  });

  if (error) {
    console.error('Error inserting matchup log:', error);
  }

  return loggedAt;
}
