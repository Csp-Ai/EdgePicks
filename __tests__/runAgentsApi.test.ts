/** @jest-environment node */
import handler from '../pages/api/run-agents';
import { runFlow } from '../lib/flow/runFlow';
import { logToSupabase } from '../lib/logToSupabase';

jest.mock('../lib/flow/runFlow');
jest.mock('../lib/logToSupabase');

const mockedRunFlow = runFlow as jest.Mock;
const mockedLog = logToSupabase as jest.Mock;

describe('run-agents API', () => {
  it('streams agent results and logs summary', async () => {
    mockedRunFlow.mockResolvedValue({
      outputs: {
        injuryScout: { team: 'A', score: 0.7, reason: 'healthy' },
      },
      executions: [
        {
          name: 'injuryScout',
          result: { team: 'A', score: 0.7, reason: 'healthy' },
        },
      ],
    });

    const req: any = { query: { homeTeam: 'A', awayTeam: 'B', week: '1' } };
    const chunks: string[] = [];
    const res: any = {
      setHeader: jest.fn(),
      write: (c: string) => chunks.push(c),
      end: jest.fn(),
      flush: jest.fn(),
      flushHeaders: jest.fn(),
    };

    await handler(req, res);

    expect(mockedRunFlow).toHaveBeenCalled();
    expect(mockedLog).toHaveBeenCalled();
    const summary = chunks.find((c) => c.includes('summary'));
    expect(summary).toBeDefined();
  });

  it('matches snapshot of agent stream', async () => {
    mockedRunFlow.mockResolvedValue({
      outputs: {
        injuryScout: { team: 'A', score: 0.72, reason: 'Key WR out' },
      },
      executions: [
        {
          name: 'injuryScout',
          result: { team: 'A', score: 0.72, reason: 'Key WR out' },
        },
      ],
    });

    const req: any = { query: { homeTeam: 'A', awayTeam: 'B', week: '1' } };
    const chunks: string[] = [];
    const res: any = {
      setHeader: jest.fn(),
      write: (c: string) => chunks.push(c),
      end: jest.fn(),
      flush: jest.fn(),
      flushHeaders: jest.fn(),
    };

    await handler(req, res);

    expect(chunks.join('')).toMatchSnapshot();
  });
});
