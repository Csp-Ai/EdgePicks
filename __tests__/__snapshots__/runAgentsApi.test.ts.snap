// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`run-agents API matches snapshot of agent stream 1`] = `
"data: {"type":"summary","sessionId":"test-session","matchup":{"homeTeam":"A","awayTeam":"B","matchDay":1,"time":"","league":""},"agents":{"injuryScout":{"team":"A","score":0.72,"reason":"Key WR out"}},"pick":{"winner":"A","confidence":0.36,"topReasons":["Key WR out"]},"cacheVersion":"v1"}

"
`;
